### PR TITLE
[OPPRO-261] Support DPP on Velox backend

### PR DESCRIPTION
**Pull request by rui-mo**
## What changes were proposed in this pull request?

Support reading partitioned table with Velox data reader.
Supported c2r in VeloxBuildSideRelation.
Supported reuse.

## How was this patch tested?

Verified on TPC-DS partitioned table.

### CODE REVIEW

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 1]
```diff
@@ -530,8 +530,19 @@ class VeloxPlanConverter::WholeStageResIterFirstStage
       std::vector<std::shared_ptr<ConnectorSplit>> connectorSplits;
       connectorSplits.reserve(paths.size());
       for (int idx = 0; idx < paths.size(); idx++) {
+        auto partitionInfo = extractPartitionColumnAndValue(paths[idx]);
+        std::unordered_map<std::string, std::optional<std::string>>
```
Comment (jinchengchenghh) [inline cpp/velox/compute/VeloxPlanConverter.cc:534]: partitionKeys should be out of for loop, is that?

Comment (rui-mo) [reply cpp/velox/compute/VeloxPlanConverter.cc:534]: Actually the map is only for a single spilt, and split path is sth like below:
```
/mnt/DP_disk1/tpcds_parquet_part_sf100/catalog_returns/cr_returned_date_sk=2451190/part-00120-f24ce1af-4ef3-44d4-9432-c996c5d48970.c000.snappy.parquet
```
What I'm doing here is to extract the partition column and value from `cr_returned_date_sk=2451190` and let the hive connector split know its partition column/value.

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 2]
```diff
@@ -587,6 +598,25 @@ class VeloxPlanConverter::WholeStageResIterFirstStage
   std::vector<core::PlanNodeId> streamIds_;
   std::vector<std::vector<exec::Split>> splits_;
   bool noMoreSplits_ = false;
+
+  // Extract the partition column and value from a path of split.
+  // Extraction for multiple partition columns is not supported.
+  std::pair<std::string, std::string> extractPartitionColumnAndValue(
+      const std::string& filePath) {
+    std::pair<std::string, std::string> partitionColumnAndValue("", "");
+    std::string delimiter = "=";
```
Comment (jinchengchenghh) [inline cpp/velox/compute/VeloxPlanConverter.cc:607]: sometimes `=` can be in column name, but this is a corner case. we do not need to consider now

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 3]
```diff
@@ -587,6 +598,25 @@ class VeloxPlanConverter::WholeStageResIterFirstStage
   std::vector<core::PlanNodeId> streamIds_;
   std::vector<std::vector<exec::Split>> splits_;
   bool noMoreSplits_ = false;
+
+  // Extract the partition column and value from a path of split.
+  // Extraction for multiple partition columns is not supported.
```
Comment (jinchengchenghh) [inline cpp/velox/compute/VeloxPlanConverter.cc:603]: Multiple partition columns is a common scene, it's better to throw unsupported exception. What will happen if we meet this case now?

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 1 above)
Comment (jinchengchenghh) [reply cpp/velox/compute/VeloxPlanConverter.cc:534]: ok

Comment (jinchengchenghh) [inline cpp/velox/compute/VeloxPlanConverter.cc:534]: Sometimes partition key with same column name can take place several times, path is like `a=2012/a=2`, the original propose is not to support this case?

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 3 above)
Comment (rui-mo) [reply cpp/velox/compute/VeloxPlanConverter.cc:603]: I guess the case you refer to is like `/my_dataset/year=2022/month=July`. I can support this case as weill.

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 1 above)
Comment (rui-mo) [reply cpp/velox/compute/VeloxPlanConverter.cc:534]: When will several partition values exist for a single split, could you tell me more about that?

Comment (jinchengchenghh) [reply cpp/velox/compute/VeloxPlanConverter.cc:534]: As in https://iceberg.apache.org/docs/latest/spark-ddl/, it has syntax `CREATE TABLE prod.db.sample (
    id bigint,
    data string,
    category string,
    ts timestamp)
USING iceberg
PARTITIONED BY (years(ts), days(ts), category)`
partition by can be the same column.

I reconsider this case, the path will be ts_years=x/ts_days=x，so there is no possible to exists this case

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 2 above)
Comment (rui-mo) [reply cpp/velox/compute/VeloxPlanConverter.cc:604]: Added a comment for this case.

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 3 above)
Comment (rui-mo) [reply cpp/velox/compute/VeloxPlanConverter.cc]: Fixed.

**Review by jinchengchenghh — APPROVED** [review]

**Review by zzcclp — APPROVED** [review]
LGTM

**Comment by rui-mo** [comment]
@zhouyuan do you have any further comment?

**Pull request merged**